### PR TITLE
HDDS-1969. Implement OM GetDelegationToken request to use Cache and DoubleBuffer.

### PR DESCRIPTION
**Pull request by bharatviswa504**

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 43 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 14 | Maven dependency ordering for branch |
| +1 | mvninstall | 602 | trunk passed |
| +1 | compile | 365 | trunk passed |
| +1 | checkstyle | 62 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 836 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 149 | trunk passed |
| 0 | spotbugs | 420 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 616 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 27 | Maven dependency ordering for patch |
| +1 | mvninstall | 566 | the patch passed |
| +1 | compile | 367 | the patch passed |
| +1 | cc | 367 | the patch passed |
| +1 | javac | 367 | the patch passed |
| +1 | checkstyle | 71 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 665 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 159 | the patch passed |
| +1 | findbugs | 638 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 284 | hadoop-hdds in the patch passed. |
| -1 | unit | 1653 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 52 | The patch does not generate ASF License warnings. |
| | | 7359 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.client.rpc.TestWatchForCommit |
|   | hadoop.ozone.client.rpc.TestOzoneAtRestEncryption |
|   | hadoop.ozone.client.rpc.TestOzoneClientRetriesOnException |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1296/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1296 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle cc |
| uname | Linux b85568b6c30b 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 167acd8 |
| Default Java | 1.8.0_222 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1296/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1296/1/testReport/ |
| Max. process+thread count | 5403 (vs. ulimit of 5500) |
| modules | C: hadoop-ozone/common hadoop-ozone/ozone-manager U: hadoop-ozone |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1296/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 55 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 13 | Maven dependency ordering for branch |
| +1 | mvninstall | 586 | trunk passed |
| +1 | compile | 350 | trunk passed |
| +1 | checkstyle | 59 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 800 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 154 | trunk passed |
| 0 | spotbugs | 413 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 616 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 16 | Maven dependency ordering for patch |
| +1 | mvninstall | 532 | the patch passed |
| +1 | compile | 366 | the patch passed |
| +1 | cc | 366 | the patch passed |
| +1 | javac | 366 | the patch passed |
| +1 | checkstyle | 73 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 638 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 149 | the patch passed |
| +1 | findbugs | 626 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 302 | hadoop-hdds in the patch passed. |
| -1 | unit | 1841 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 39 | The patch does not generate ASF License warnings. |
| | | 7361 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.om.TestScmSafeMode |
|   | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1296/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1296 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle cc |
| uname | Linux 00a46a5c70c6 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 9b8359b |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1296/2/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1296/2/testReport/ |
| Max. process+thread count | 4832 (vs. ulimit of 5500) |
| modules | C: hadoop-ozone/common hadoop-ozone/ozone-manager U: hadoop-ozone |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1296/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-ozone/common/src/main/proto/OzoneManagerProtocol.proto [hunk 1]
```diff
@@ -307,6 +308,13 @@ message UserInfo {
     optional string remoteAddress = 3;
 }
 
+/**
+  This will be used during OM HA, once leader generates token send's this
```
Comment (arp7) [inline hadoop-ozone/common/src/main/proto/OzoneManagerProtocol.proto]: Can you expand on this comment a bit?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/security/OMGetDelegationTokenRequest.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+    // Call OM to create token
+    Token<OzoneTokenIdentifier> token = ozoneManager
+        .getDelegationToken(new Text(getDelegationTokenRequest.getRenewer()));
+
+    // Create UpdateGetDelegationTokenRequest with token response.
+    OMRequest.Builder omRequest = OMRequest.newBuilder()
+        .setUserInfo(getUserInfo())
+        .setUpdateGetDelegationTokenRequest(
+            UpdateGetDelegationTokenRequest.newBuilder()
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/security/OMGetDelegationTokenRequest.java]: Can you add some explanation of what is going on here... same that we discussed offline.

**Comment by bharatviswa504** [comment]
Thank You @arp7 for the review.
Fixed review comments.

**Comment by bharatviswa504** [comment]
Thank You @arp7 for the review.
I will commit this to the trunk. 2nd PR changes are only code comment changes. Committing this, as previous CI test failures are not related to this PR.

**Pull request merged**